android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Node.kt

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (C) 2017. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.badoo.ribs.core
+
+import android.os.Bundle
+import android.os.Parcelable
+import android.support.annotation.CallSuper
+import android.support.annotation.MainThread
+import android.support.annotation.VisibleForTesting
+import android.util.SparseArray
+import android.view.ViewGroup
+import com.badoo.ribs.core.view.RibView
+import com.badoo.ribs.core.view.ViewFactory
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.uber.rib.util.RibRefWatcher
+import java.util.UUID
+import java.util.concurrent.CopyOnWriteArrayList
+
+/**
+ * Responsible for handling the addition and removal of child nodes.
+ **/
+open class Node<V : RibView>(
+    val forClass: Class<*>,
+    private val viewFactory: ViewFactory<V>?,
+    private val router: Router<*, V>,
+    private val interactor: Interactor<*, V>,
+    private val ribRefWatcher: RibRefWatcher = RibRefWatcher.getInstance()
+) {
+    companion object {
+        internal const val KEY_CHILD_NODES = "node.children"
+        internal const val KEY_ROUTER = "node.router"
+        internal const val KEY_INTERACTOR = "node.interactor"
+        private const val KEY_RIB_ID = "rib.id"
+        private const val KEY_VIEW_STATE = "view.state"
+        private val requestCodeRegistry = RequestCodeRegistry(8)
+    }
+
+    init {
+        router.node = this
+    }
+
+    private var savedInstanceState: Bundle? = null
+    internal val children = CopyOnWriteArrayList<Node<*>>()
```

Comment:
maybe childs?
usually reader expect to see 's at the end if this field hold multiple references.

Reply:
Children is a correct plural for child

Reply:
I understand that children is correct, I am talking that it is more natural for you to see 's at the end when you are reading code, if variable contains references to multiple objects.

Reply:
We can say `childrenNodes` ?